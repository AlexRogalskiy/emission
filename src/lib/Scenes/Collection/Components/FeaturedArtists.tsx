import { Box, Button, EntityHeader, Flex, Sans } from "@artsy/palette"
import { FeaturedArtists_collection } from "__generated__/FeaturedArtists_collection.graphql"
import SwitchBoard from "lib/NativeModules/SwitchBoard"
import { get } from "lib/utils/get"
import React from "react"
import { TouchableHighlight, TouchableWithoutFeedback } from "react-native"
import { createFragmentContainer, graphql } from "react-relay"

interface FeaturedArtistsProps {
  collection: FeaturedArtists_collection
}

interface FeaturedArtistsState {
  showMore: boolean
}

export class FeaturedArtists extends React.Component<FeaturedArtistsProps, FeaturedArtistsState> {
  state = {
    showMore: false,
  }

  handleTap = (context: any, href: string) => {
    SwitchBoard.presentNavigationViewController(context, href)
  }

  getFeaturedArtistEntityCollection = (
    artists: FeaturedArtists_collection["artworksConnection"]["merchandisableArtists"]
  ) => {
    return artists.map((artist, index) => {
      const hasArtistMetaData = artist.nationality && artist.birthday
      return (
        <Box width="100%" key={index} pb={20}>
          <TouchableWithoutFeedback onPress={() => this.handleTap(this, `/artist/${artist.slug}`)}>
            <EntityHeader
              imageUrl={artist.imageUrl}
              name={artist.name}
              meta={hasArtistMetaData ? `${artist.nationality}, b. ${artist.birthday}` : undefined}
              href={`/artist/${artist.slug}`}
              FollowButton={
                <Button
                  variant={artist.isFollowed ? "primaryBlack" : "secondaryOutline"}
                  size="small"
                  longestText="Following"
                >
                  {artist.isFollowed ? "Following" : "Follow"}
                </Button>
              }
            />
          </TouchableWithoutFeedback>
        </Box>
      )
    })
  }

  render() {
    const artists = get(this.props, p => p.collection.artworksConnection.merchandisableArtists)
    if (!artists) {
      return null
    }

    const hasMultipleArtists = artists.length > 1

    const artistCount = 3
    const remainingCount = artists.length - artistCount
    const truncatedArtists = this.getFeaturedArtistEntityCollection(artists).slice(0, artistCount)
    const headlineLabel = "Featured Artist" + (hasMultipleArtists ? "s" : "")

    return (
      <Box pb={1}>
        <Sans size="2" weight="medium" pb={15}>
          {headlineLabel}
        </Sans>
        <Flex flexWrap="wrap">
          {this.state.showMore || artists.length <= artistCount ? (
            this.getFeaturedArtistEntityCollection(artists)
          ) : (
            <>
              {truncatedArtists}
              <TouchableHighlight
                onPress={() => {
                  this.setState({ showMore: true })
                }}
              >
                <EntityHeader initials={`+ ${remainingCount}`} name="View more" />
              </TouchableHighlight>
            </>
          )}
        </Flex>
      </Box>
    )
  }
}

export const CollectionFeaturedArtistsContainer = createFragmentContainer(FeaturedArtists, {
  collection: graphql`
    fragment FeaturedArtists_collection on MarketingCollection {
      # TODO: size:9 is not actually limiting to 9 items. We need to figure out
      #  why the back-end is not respecting that argument.
      artworksConnection(aggregations: [MERCHANDISABLE_ARTISTS], size: 9, sort: "-decayed_merch") {
        merchandisableArtists {
          slug
          internalID
          name
          imageUrl
          birthday
          nationality
          isFollowed
        }
      }
    }
  `,
})
