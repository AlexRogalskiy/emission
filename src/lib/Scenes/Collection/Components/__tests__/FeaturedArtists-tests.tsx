import { Theme } from "@artsy/palette"
import { FeaturedArtists_collection } from "__generated__/FeaturedArtists_collection.graphql"
import { renderRelayTree } from "lib/tests/renderRelayTree"
import React from "react"
import { graphql } from "react-relay"
import { CollectionFeaturedArtistsContainer as FeaturedArtists } from "../FeaturedArtists"

jest.unmock("react-relay")

const CollectionFixture: FeaturedArtists_collection = {
  artworksConnection: {
    merchandisableArtists: [
      {
        slug: "pablo-picasso",
        internalID: "1234",
        name: "Pablo Picasso",
        image: {
          resized: {
            url: "/some/resized/picasso/image/url",
          },
        },
        birthday: "1877",
        nationality: "American",
        isFollowed: true,
      },
      {
        slug: "andy-warhol",
        internalID: "2342",
        name: "Andy Warhol",
        image: {
          resized: {
            url: "/some/resized/warhol/image/url",
          },
        },
        birthday: "1947",
        nationality: "American",
        isFollowed: false,
      },
      {
        slug: "joan-miro",
        internalID: "3454",
        name: "Joan Miro",
        image: {
          resized: {
            url: "/some/resized/miro/image/url",
          },
        },
        birthday: "1877",
        nationality: "Spanish",
        isFollowed: true,
      },
    ],
  },
  " $refType": null,
}

describe("FeaturedArtists", () => {
  const render = () =>
    renderRelayTree({
      Component: ({ marketingCollection }) => (
        <Theme>
          <FeaturedArtists collection={marketingCollection} />
        </Theme>
      ),
      query: graphql`
        query FeaturedArtistsTestsQuery @raw_response_type {
          marketingCollection(slug: "emerging-photographers") {
            ...FeaturedArtists_collection
          }
        }
      `,
      mockData: {
        marketingCollection: CollectionFixture,
      },
    })

  it("renders properly", async () => {
    const tree = await render()
    expect(tree.html()).toMatchSnapshot()
  })

  it("renders an EntityHeader for each featured artist", async () => {
    const tree = await render()

    const entityHeaders = tree.find("EntityHeader")
    expect(entityHeaders.length).toEqual(3)

    const output = tree.html()
    expect(output).toContain("Pablo Picasso")
    expect(output).toContain("Andy Warhol")
    expect(output).toContain("Joan Miro")
  })
})
