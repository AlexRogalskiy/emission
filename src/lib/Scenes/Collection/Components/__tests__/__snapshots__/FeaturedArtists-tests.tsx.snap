// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeaturedArtists renders properly 1`] = `"<view pb=\\"1\\"><text font-family=\\"Unica77LL-Medium\\" font-size=\\"12px\\" lineheight=\\"16px\\" pb=\\"15\\">Featured Artists</text><view flexwrap=\\"wrap\\"><view width=\\"100%\\" pb=\\"20\\"><view flexdirection=\\"row\\" justifycontent=\\"space-between\\" flexwrap=\\"nowrap\\" focusable=\\"true\\"><view mr=\\"1\\" justifycontent=\\"center\\"><image resizemode=\\"stretch\\" style=\\"width: 45px; height: 45px; border-radius: 22.5px;\\" source=\\"[object Object]\\"></image></view><view justifycontent=\\"center\\" width=\\"0\\" flexgrow=\\"1\\"><text font-family=\\"ReactNativeAGaramondPro-Regular\\" font-size=\\"16px\\" lineheight=\\"20px\\" ellipsizemode=\\"tail\\" numberoflines=\\"1\\" color=\\"black100\\">Pablo Picasso</text><text font-family=\\"Unica77LL-Regular\\" font-size=\\"12px\\" lineheight=\\"16px\\" ellipsizemode=\\"tail\\" numberoflines=\\"1\\" color=\\"black60\\">American, b. 1877</text></view><view ml=\\"1\\" flexdirection=\\"row\\" alignitems=\\"center\\" justifycontent=\\"flex-end\\"><view flexdirection=\\"row\\" focusable=\\"true\\"><view variant=\\"primaryBlack\\" theme=\\"[object Object]\\" px=\\"1\\"><view><text font-family=\\"Unica77LL-Medium\\" font-size=\\"12px\\" lineheight=\\"16px\\" color=\\"#FFF\\">Following</text></view><text font-family=\\"Unica77LL-Medium\\" font-size=\\"12px\\" lineheight=\\"16px\\" role=\\"presentation\\">Following</text></view></view></view></view></view><view width=\\"100%\\" pb=\\"20\\"><view flexdirection=\\"row\\" justifycontent=\\"space-between\\" flexwrap=\\"nowrap\\" focusable=\\"true\\"><view mr=\\"1\\" justifycontent=\\"center\\"><image resizemode=\\"stretch\\" style=\\"width: 45px; height: 45px; border-radius: 22.5px;\\" source=\\"[object Object]\\"></image></view><view justifycontent=\\"center\\" width=\\"0\\" flexgrow=\\"1\\"><text font-family=\\"ReactNativeAGaramondPro-Regular\\" font-size=\\"16px\\" lineheight=\\"20px\\" ellipsizemode=\\"tail\\" numberoflines=\\"1\\" color=\\"black100\\">Andy Warhol</text><text font-family=\\"Unica77LL-Regular\\" font-size=\\"12px\\" lineheight=\\"16px\\" ellipsizemode=\\"tail\\" numberoflines=\\"1\\" color=\\"black60\\">American, b. 1947</text></view><view ml=\\"1\\" flexdirection=\\"row\\" alignitems=\\"center\\" justifycontent=\\"flex-end\\"><view flexdirection=\\"row\\" focusable=\\"true\\"><view variant=\\"secondaryOutline\\" theme=\\"[object Object]\\" px=\\"1\\"><view><text font-family=\\"Unica77LL-Medium\\" font-size=\\"12px\\" lineheight=\\"16px\\" color=\\"#000\\">Follow</text></view><text font-family=\\"Unica77LL-Medium\\" font-size=\\"12px\\" lineheight=\\"16px\\" role=\\"presentation\\">Following</text></view></view></view></view></view><view width=\\"100%\\" pb=\\"20\\"><view flexdirection=\\"row\\" justifycontent=\\"space-between\\" flexwrap=\\"nowrap\\" focusable=\\"true\\"><view mr=\\"1\\" justifycontent=\\"center\\"><image resizemode=\\"stretch\\" style=\\"width: 45px; height: 45px; border-radius: 22.5px;\\" source=\\"[object Object]\\"></image></view><view justifycontent=\\"center\\" width=\\"0\\" flexgrow=\\"1\\"><text font-family=\\"ReactNativeAGaramondPro-Regular\\" font-size=\\"16px\\" lineheight=\\"20px\\" ellipsizemode=\\"tail\\" numberoflines=\\"1\\" color=\\"black100\\">Joan Miro</text><text font-family=\\"Unica77LL-Regular\\" font-size=\\"12px\\" lineheight=\\"16px\\" ellipsizemode=\\"tail\\" numberoflines=\\"1\\" color=\\"black60\\">Spanish, b. 1877</text></view><view ml=\\"1\\" flexdirection=\\"row\\" alignitems=\\"center\\" justifycontent=\\"flex-end\\"><view flexdirection=\\"row\\" focusable=\\"true\\"><view variant=\\"primaryBlack\\" theme=\\"[object Object]\\" px=\\"1\\"><view><text font-family=\\"Unica77LL-Medium\\" font-size=\\"12px\\" lineheight=\\"16px\\" color=\\"#FFF\\">Following</text></view><text font-family=\\"Unica77LL-Medium\\" font-size=\\"12px\\" lineheight=\\"16px\\" role=\\"presentation\\">Following</text></view></view></view></view></view></view></view>"`;
